csharp/src/Ice/Reference.cs

```diff
@@ -238,892 +338,512 @@ public override string ToString()
+                            }
+                            endpoint = endpoint.ConnectionId(connectionId);
+                            if (connectionTimeout != null)
+                            {
+                                endpoint = endpoint.Timeout(connectionTimeout.Value);
+                            }
+                            return endpoint;
+                        }).ToArray();
                     }
                 }
-                // else all good already
-            }
-            else if (IsConnectionCached != reference.IsConnectionCached)
-            {
-                Debug.Assert(!reference.IsConnectionCached); // it's a freshly cloned reference
-                reference._requestHandlerMutex = new object();
+
+                var clone = new Reference(adapterId ?? AdapterId,
+                                      cacheConnection ?? IsConnectionCached,
+                                      collocationOptimized ?? IsCollocationOptimized,
+                                      Communicator,
+                                      compress ?? Compress,
+                                      connectionId ?? ConnectionId,
+                                      connectionTimeout ?? ConnectionTimeout,
+                                      context ?? Context,
+                                      encoding ?? Encoding,
+                                      endpointSelection ?? EndpointSelection,
+                                      endpoints ?? Endpoints,
+                                      facet ?? Facet,
+                                      identity ?? Identity,
+                                      invocationMode ?? InvocationMode,
+                                      invocationTimeout ?? InvocationTimeout,
+                                      locatorCacheTimeout ?? LocatorCacheTimeout,
+                                      locatorInfo, // no fallback otherwise breaks clearLocator
+                                      preferSecure ?? IsPreferSecure,
+                                      protocol ?? Protocol,
+                                      routerInfo, // no fallback otherwise breaks clearRouter
+                                      secure ?? IsSecure);
+
+                return clone == this ? this : clone;
             }
-            return reference;
         }
 
-        public override void StreamWrite(OutputStream s)
+        // Marshal the reference.
+        internal void StreamWrite(OutputStream ostr)
         {
-            base.StreamWrite(s);
+            if (IsFixed)
+            {
+                throw new NotSupportedException("cannot marshal a fixed proxy");
+            }
 
-            s.WriteSize(_endpoints.Length);
-            if (_endpoints.Length > 0)
+            // Don't write the identity here. Operations calling StreamWrite write the identity.
+            // TODO: why not write identity here?
```

Comment:
I don't know, writing it here would make a lot more sense, we should fix this.

Reply:
Fixed.